crates/red_knot_test/src/assertion.rs

```diff
@@ -0,0 +1,621 @@
+    type Item = LineAssertions<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let file = self.file_assertions;
+        let ranged_assertion = self.inner.next()?;
+        let mut collector = AssertionVec::new();
+        let mut line = file.line_number(&ranged_assertion);
+        // Collect all own-line comments on consecutive lines; these all apply to the same line of
+        // code. For example:
+        //
+        //     # Error: [unbound-name]
+        //     # Type: Unbound
+        //     reveal_type(x)
+        //
+        if file.is_own_line(&ranged_assertion) {
+            collector.push(ranged_assertion.into());
+            let mut only_own_line = true;
+            while let Some(ranged_assertion) = self.inner.peek() {
+                let next_line = line.saturating_add(1);
+                if file.line_number(ranged_assertion) == next_line {
+                    if !file.is_own_line(ranged_assertion) {
+                        only_own_line = false;
+                    }
+                    line = next_line;
+                    collector.push(self.inner.next().unwrap().into());
+                    // If we see an end-of-line comment, it has to be the end of the stack,
+                    // otherwise we'd botch this case, attributing all three errors to the `bar`
+                    // line:
+                    //
+                    //     # Error:
+                    //     foo  # Error:
+                    //     bar  # Error:
+                    if !only_own_line {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+            if only_own_line {
+                // The collected comments apply to the _next_ line in the code.
+                line = line.saturating_add(1);
+            }
+        } else {
+            // We have a line-trailing comment; it applies to its own line, and is not grouped.
+            collector.push(ranged_assertion.into());
+        }
+        Some(LineAssertions {
+            line,
+            assertions: collector,
+        })
+    }
+}
+
+impl std::iter::FusedIterator for LineAssertionsIterator<'_> {}
+
+/// One or more assertions referring to the same line of code.
+#[derive(Debug)]
+pub(crate) struct LineAssertions<'a> {
+    /// The line these assertions refer to.
+    ///
+    /// Not necessarily the same line the assertion comment is located on; for an own-line comment,
+    /// it's the next non-assertion line.
+    pub(crate) line: OneIndexed,
+
+    /// The assertions referring to this line.
+    pub(crate) assertions: AssertionVec<'a>,
+}
+
+impl<'a, 'b> From<&'a LineAssertions<'b>> for &'a [Assertion<'b>] {
+    fn from(value: &'a LineAssertions<'b>) -> Self {
+        &value.assertions[..]
+    }
+}
+
+static TYPE_RE: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^#\s*Type:\s*(?<ty_display>.+?)\s*$").unwrap());
```

Comment:
Hmm. I definitely agree that we should prefer and use the spaced style. It just feels like a confusing/bad experience to silently not treat the comment as an assertion because you left out a space.

If we want the test framework to enforce an opinionated style, I think I'd rather still regex-match leniently, and then issue an explicit error about the lack of spaces. But then I'm not sure that's a priority right now either :)

Reply:
I don't object to adding a feature to error if there are missing spaces in an assertion comment, but I don't want to do it in this PR. If someone is motivated to do it, or we have trouble enforcing our preferred style without it, we can do it as its own PR.